include/deal.II/multigrid/multigrid.h

```diff
@@ -34,29 +34,77 @@ DEAL_II_NAMESPACE_OPEN
 /*!@addtogroup mg */
 /*@{*/
 
+
 /**
- * Implementation of the multigrid method. The implementation supports both
- * continuous and discontinuous elements and follows the procedure described in
- * the @ref mg_paper "multigrid paper by Janssen and Kanschat".
- *
- * The function which starts a multigrid cycle on the finest level is cycle().
- * Depending on the cycle type chosen with the constructor (see enum Cycle),
- * this function triggers one of the cycles level_v_step() or level_step(),
- * where the latter one can do different types of cycles.
- *
- * Using this class, it is expected that the right hand side has been
- * converted from a vector living on the locally finest level to a multilevel
- * vector. This is a nontrivial operation, usually initiated automatically by
- * the class PreconditionMG and performed by the classes derived from
- * MGTransferBase.
- *
- * @note The interface of this class is still very clumsy. In particular, you
- * will have to set up quite a few auxiliary objects before you can use it.
- * Unfortunately, it seems that this can be avoided only be restricting the
- * flexibility of this class in an unacceptable way.
- *
- * @author Guido Kanschat, 1999 - 2005
- */
+  */
+struct Event
+{
+  enum type_t {start, end, single};
+  enum event_name_t {transfer_to_mg, transfer_to_global, coarse_solve, transfer_down_to, transfer_up_to, smoother_step};
+  type_t type;
+  event_name_t name;
+  unsigned int level;
+};
+
+inline const char *to_string(const Event::type_t t)
+{
+  switch (t)
+    {
+    case Event::start:
+      return "start";
+    case Event::end:
+      return "end";
+    case Event::single:
+      return "single";
+    }
+  Assert(false, ExcNotImplemented());
+  return "";
+}
+
+inline const char *to_string(const Event::event_name_t n)
+{
+  switch (n)
+    {
+    case Event::transfer_to_mg:
+      return "transfer_to_mg";
+    case Event::transfer_to_global:
+      return "transfer_to_global";
+    case Event::coarse_solve:
+      return "coarse_solve";
+    case Event::transfer_down_to:
+      return "transfer_down_to";
+    case Event::transfer_up_to:
+      return "transfer_up_to";
+    case Event::smoother_step:
+      return "smoother_step";
+    }
+  Assert(false, ExcNotImplemented());
+  return "";
+}
+
+/**
+* Implementation of the multigrid method. The implementation supports both
```

Comment:
you probably did not need to indent this comment so that `git` does not show it as a diff.